mllib/src/main/scala/org/apache/spark/ml/reduction/OneVsAll.scala

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.reduction
+
+import scala.language.existentials
+import scala.reflect.ClassTag
+
+import org.apache.spark.annotation.{AlphaComponent, Experimental}
+import org.apache.spark.ml.attribute.BinaryAttribute
+import org.apache.spark.ml.classification.{ClassificationModel, Classifier, ClassifierParams}
+import org.apache.spark.ml.impl.estimator.{PredictionModel, Predictor}
+import org.apache.spark.ml.param.ParamMap
+import org.apache.spark.ml.util.{MetadataUtils, SchemaUtils}
+import org.apache.spark.mllib.linalg.Vector
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+import org.apache.spark.storage.StorageLevel
+
+/**
+ * Model produced by [[OneVsAll]].
+ *
+ * @param parent
+ * @param models the binary classification models for reduction.
+ */
+@AlphaComponent
+class OneVsAllModel[
+    FeaturesType,
+    E <: Classifier[FeaturesType, E, M],
```

Comment:
But the type parameters E and M are for OneVsAll and OneVsAllModel, which are of types Predictor and PredictionModel.  (It bothers me that this compiles, hm...)  Those type parameters should not affect what OneVsAllModel stores internally (a ClassificationModel).